File: app/src/main/java/com/ludoscity/findmybikes/helpers/FavoriteEntityStationDao.java
package com.ludoscity.findmybikes.helpers;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;

import com.ludoscity.findmybikes.datamodel.FavoriteEntityStation;

import java.util.List;

/**
 * Created by F8Full on 2017-12-23. This file is part of #findmybikes
 */
@Dao
public interface FavoriteEntityStationDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertOne(FavoriteEntityStation favoriteEntityStation);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertAll(List<FavoriteEntityStation> favoriteEntityStationList);

    @Query("DELETE FROM FavoriteEntityStation WHERE id = :favoriteId")
    void deleteOne(String favoriteId);

    @Query("SELECT * FROM FavoriteEntityStation ORDER BY ui_index DESC")
    LiveData<List<FavoriteEntityStation>> getAll();

    //TODO: a more complex query that can be returned as a LiveData<FavoriteEntityBase> ?
    @Query("SELECT * FROM FavoriteEntityStation WHERE id = :favoriteId")
    LiveData<FavoriteEntityStation> getForId(String favoriteId);

    //TODO: this seems broken
    @Query("SELECT COUNT(*) FROM FavoriteEntityStation WHERE id <> :favoriteId")
    LiveData<Long> validFavoriteCount(String favoriteId);

    @Query("DELETE FROM favoriteentitystation")
    void deleteAll();

}

File: app/src/main/java/com/ludoscity/findmybikes/data/database/favorite/FavoriteEntityStationDao.kt
package com.ludoscity.findmybikes.data.database.favorite

import android.arch.lifecycle.LiveData
import android.arch.persistence.room.Dao
import android.arch.persistence.room.Insert
import android.arch.persistence.room.OnConflictStrategy
import android.arch.persistence.room.Query

/**
 * Created by F8Full on 2017-12-23. This file is part of #findmybikes
 */
@Dao
interface FavoriteEntityStationDao {

    @Query("SELECT * FROM FavoriteEntityStation WHERE bike_system_id = :bikeSystemId ORDER BY ui_index ASC")
    fun getAllForBikeSystemId(bikeSystemId: String): LiveData<List<FavoriteEntityStation>>

    @get:Query("SELECT id FROM FavoriteEntityStation")
    val allId: List<String>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertOne(favoriteEntityStation: FavoriteEntityStation)

    @Query("UPDATE FavoriteEntityStation SET custom_name = :newCustomName WHERE id = :favoriteIdToUpdate")
    fun updateCustomNameByFavoriteId(favoriteIdToUpdate: String, newCustomName: String)

    @Query("UPDATE FavoriteEntityStation SET ui_index = :newUiIndex WHERE id = :favoriteIdToUpdate")
    fun updateUiIndexByFavoriteId(favoriteIdToUpdate: String, newUiIndex: Int)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertAll(favoriteEntityStationList: List<FavoriteEntityStation>)

    @Query("DELETE FROM FavoriteEntityStation WHERE id = :favoriteId")
    fun deleteOne(favoriteId: String)

    //TODO: a more complex query that can be returned as a LiveData<FavoriteEntityBase> ?
    @Query("SELECT * FROM FavoriteEntityStation WHERE id = :favoriteId")
    fun getForId(favoriteId: String): FavoriteEntityStation?

    @get:Query("SELECT COUNT(*) FROM FavoriteEntityStation")
    val count: Int

    @Query("SELECT COUNT(*) FROM FavoriteEntityStation WHERE id <> :favoriteIdToExclude")
    fun validFavoriteCount(favoriteIdToExclude: String): Int

    @Query("SELECT COUNT(*) FROM FavoriteEntityStation WHERE id = :favoriteIdToCheck")
    fun isFavoriteId(favoriteIdToCheck: String): Int

    /*@Query("DELETE FROM favoriteentitystation")
    void deleteAll();*/

}
